File: 0077-combinations/0077-combinations.cpp
class Solution {
private:
vector<vector<int>> result;
vector<int> combo; // individual combo size = k
void backtrack(int n, int k, int startIdx){
    // stopping criteria
    if (combo.size() == k){
        result.push_back(combo);
        return;
    }
    // single layer loop logic
    for (int i = startIdx; i<=n; i++){
        combo.push_back(i);
        backtrack(n, k, i + 1); // recurse
        combo.pop_back(); // backtrack
    }
}    

public:
    vector<vector<int>> combine(int n, int k) {
        result.clear();
        combo.clear();
        backtrack(n, k, 1);
        return result;
    }
};

File: 0077-combinations/0077-combinations.py
class Solution(object):
    def __init__(self):
        self.result = []
    
    def backtrack(self, n, k, startIdx, combo):
        if len(combo) == k:
            self.result.append(combo[:])
            return
        for i in range(startIdx, n + 1):
            combo.append(i)
            self.backtrack(n, k, i + 1, combo)
            combo.pop()

    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        self.backtrack(n, k, 1, [])
        return self.result